File: backend/app/schemas/ivr.py
"""
IVR Flow schemas.
"""
from datetime import datetime
from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field

from app.models.ivr import IVRFlowStatus, IVRNodeType


# =============================================================================
# Node Schemas
# =============================================================================

class IVRNodePosition(BaseModel):
    """Position of a node in the flow editor."""
    x: float
    y: float


class IVRNodeData(BaseModel):
    """Data for an IVR node. Structure depends on node type."""
    # Common fields
    label: Optional[str] = None

    # Audio-related
    audio_file_id: Optional[str] = None
    prompt_audio_id: Optional[str] = None

    # Menu options (key = DTMF digit, value = target node ID)
    options: Optional[Dict[str, str]] = None
    timeout: Optional[int] = Field(None, ge=1, le=60)
    max_retries: Optional[int] = Field(None, ge=1, le=10)

    # Survey question
    question_id: Optional[str] = None
    valid_inputs: Optional[List[str]] = None

    # Transfer
    transfer_number: Optional[str] = None

    # Conditional
    condition: Optional[str] = None
    true_target: Optional[str] = None
    false_target: Optional[str] = None

    # Variable
    variable_name: Optional[str] = None
    variable_value: Optional[str] = None

    # Wait for DTMF after playing audio
    wait_for_dtmf: Optional[bool] = False

    class Config:
        extra = "allow"  # Allow additional fields for extensibility


class IVRNode(BaseModel):
    """An IVR flow node."""
    id: str = Field(..., description="Unique node identifier")
    type: IVRNodeType = Field(..., description="Node type")
    position: IVRNodePosition = Field(..., description="Position in flow editor")
    data: IVRNodeData = Field(default_factory=IVRNodeData)


class IVREdge(BaseModel):
    """An edge connecting two nodes."""
    id: Optional[str] = None
    source: str = Field(..., description="Source node ID")
    target: str = Field(..., description="Target node ID")
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None


class IVRFlowDefinition(BaseModel):
    """Complete IVR flow definition."""
    nodes: List[IVRNode] = Field(default_factory=list)
    edges: List[IVREdge] = Field(default_factory=list)
    start_node: Optional[str] = None


class IVRViewport(BaseModel):
    """React Flow viewport state."""
    x: float = 0
    y: float = 0
    zoom: float = 1


# =============================================================================
# Version Schemas
# =============================================================================

class IVRFlowVersionBase(BaseModel):
    """Base version schema."""
    notes: Optional[str] = Field(None, max_length=1000)


class IVRFlowVersionCreate(IVRFlowVersionBase):
    """Schema for creating a new version."""
    definition: IVRFlowDefinition = Field(..., description="Flow definition")
    viewport: Optional[IVRViewport] = None


class IVRFlowVersionResponse(IVRFlowVersionBase):
    """Version response schema."""
    id: str
    flow_id: str
    version: int
    definition: Dict[str, Any]  # Raw dict for flexibility
    viewport: Optional[Dict[str, Any]] = None
    created_at: datetime
    created_by_id: Optional[str] = None

    class Config:
        from_attributes = True


# =============================================================================
# Flow Schemas
# =============================================================================

class IVRFlowBase(BaseModel):
    """Base IVR flow schema."""
    name: str = Field(..., min_length=1, max_length=255, description="Flow name")
    description: Optional[str] = Field(None, max_length=1000)


class IVRFlowCreate(IVRFlowBase):
    """Schema for creating an IVR flow."""
    definition: Optional[IVRFlowDefinition] = Field(
        None,
        description="Initial flow definition (creates first version)"
    )


class IVRFlowUpdate(BaseModel):
    """Schema for updating IVR flow metadata."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)


class IVRFlowResponse(IVRFlowBase):
    """IVR flow response schema."""
    id: str
    organization_id: str
    status: IVRFlowStatus
    active_version_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by_id: Optional[str] = None

    # Include latest version info if available
    version_count: Optional[int] = None

    class Config:
        from_attributes = True


class IVRFlowDetailResponse(IVRFlowResponse):
    """Detailed IVR flow response with active version definition."""
    active_version: Optional[IVRFlowVersionResponse] = None


class IVRFlowListResponse(BaseModel):
    """Response for listing IVR flows."""
    items: List[IVRFlowResponse]
    total: int
    page: int
    page_size: int


# =============================================================================
# Action Schemas
# =============================================================================

class IVRFlowPublishRequest(BaseModel):
    """Request to publish a specific version."""
    version_id: Optional[str] = Field(
        None,
        description="Version ID to publish. If not provided, publishes the latest version."
    )


class IVRFlowPublishResponse(BaseModel):
    """Response after publishing."""
    id: str
    status: IVRFlowStatus
    active_version_id: str
    message: str


class IVRFlowDuplicateRequest(BaseModel):
    """Request to duplicate an IVR flow."""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class IVRFlowValidationResult(BaseModel):
    """Result of validating an IVR flow definition."""
    is_valid: bool
    errors: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)

File: frontend/src/types/ivr.ts
// IVR Flow Builder types
import { Node, Edge } from 'reactflow';

export type IVRNodeType =
  | 'start'
  | 'play_audio'
  | 'menu'
  | 'survey_question'
  | 'record'
  | 'transfer'
  | 'conditional'
  | 'set_variable'
  | 'hangup';

export type IVRFlowStatus = 'draft' | 'published' | 'archived';

// Base node data interface
export interface BaseNodeData {
  label?: string;
}

// Specific node data types
export interface PlayAudioNodeData extends BaseNodeData {
  audio_file_id?: string;
  audio_file_name?: string;
  wait_for_dtmf?: boolean;
}

export interface MenuNodeData extends BaseNodeData {
  prompt_audio_id?: string;
  prompt_audio_name?: string;
  timeout?: number;
  max_retries?: number;
  options?: Record<string, string>; // key -> target node id
  invalid_input_audio_id?: string;
}

export interface SurveyQuestionNodeData extends BaseNodeData {
  question_id?: string;
  question_text?: string;
  prompt_audio_id?: string;
  prompt_audio_name?: string;
  valid_inputs?: string[];
  timeout?: number;
  max_retries?: number;
}

export interface RecordNodeData extends BaseNodeData {
  max_duration_seconds?: number;
  beep_before_recording?: boolean;
  finish_on_key?: string;
  timeout_seconds?: number;
}

export interface TransferNodeData extends BaseNodeData {
  destination_number?: string;
  timeout_seconds?: number;
  caller_id?: string;
}

export interface ConditionalNodeData extends BaseNodeData {
  variable_name?: string;
  operator?: 'equals' | 'not_equals' | 'greater_than' | 'less_than' | 'contains';
  value?: string;
  true_target?: string;
  false_target?: string;
}

export interface SetVariableNodeData extends BaseNodeData {
  variable_name?: string;
  value?: string;
  value_type?: 'static' | 'dtmf_input' | 'call_variable';
}

export interface HangupNodeData extends BaseNodeData {
  reason?: string;
}

export type IVRNodeData =
  | BaseNodeData
  | PlayAudioNodeData
  | MenuNodeData
  | SurveyQuestionNodeData
  | RecordNodeData
  | TransferNodeData
  | ConditionalNodeData
  | SetVariableNodeData
  | HangupNodeData;

// React Flow node with IVR data
export interface IVRNode extends Node {
  type: IVRNodeType;
  data: IVRNodeData;
}

// Flow definition
export interface IVRFlowDefinition {
  nodes: IVRNode[];
  edges: Edge[];
  start_node?: string;
}

// Flow version
export interface IVRFlowVersion {
  id: string;
  flow_id: string;
  version: number;
  definition: IVRFlowDefinition;
  viewport?: {
    x: number;
    y: number;
    zoom: number;
  };
  notes?: string;
  created_by_id?: string;
  created_at: string;
}

// Complete flow
export interface IVRFlow {
  id: string;
  name: string;
  description?: string;
  organization_id: string;
  status: IVRFlowStatus;
  active_version_id?: string;
  created_by_id?: string;
  created_at: string;
  updated_at: string;
  versions?: IVRFlowVersion[];
}

// API request/response types
export interface IVRFlowCreate {
  name: string;
  description?: string;
}

export interface IVRFlowUpdate {
  name?: string;
  description?: string;
  status?: IVRFlowStatus;
}

export interface IVRFlowVersionCreate {
  definition: IVRFlowDefinition;
  viewport?: {
    x: number;
    y: number;
    zoom: number;
  };
  notes?: string;
}

export interface IVRFlowListResponse {
  items: IVRFlow[];
  total: number;
  page: number;
  page_size: number;
}
